### PR TITLE
refactor(tests): integrate lua-client into core and use core for functionaltests

### PR DESCRIPTION
**Pull request by bfredl**
more exciting errors for actually running more code under the sanitizers!

### CODE REVIEW

File: cmake/RunTests.cmake [hunk 1]
```diff
@@ -71,16 +71,11 @@ if(NOT DEFINED ENV{TEST_TIMEOUT} OR "$ENV{TEST_TIMEOUT}" STREQUAL "")
...
-  set(RUNNER_PRG ${NVIM_PRG} -ll ${WORKING_DIR}/test/busted_runner.lua)
-else()
-  set(RUNNER_PRG ${BUSTED_PRG})
-endif()
+set(ENV{DEPS_PREFIX} ${DEPS_PREFIX})  # used by test/busted_runner.lua on windows
+set(ENV{BUSTED_PRG} ${BUSTED_PRG})  # uuuu
 
 execute_process(
-  COMMAND ${RUNNER_PRG} -v -o test.busted.outputHandlers.${BUSTED_OUTPUT_TYPE}
+  COMMAND ${NVIM_PRG} -ll ${WORKING_DIR}/test/busted_runner.lua -v -o test.busted.outputHandlers.${BUSTED_OUTPUT_TYPE}
```
Comment (justinmk) [inline cmake/RunTests.cmake:78]:
```suggestion
  COMMAND ${NVIM_PRG} -l ${WORKING_DIR}/test/busted_runner.lua -v -o test.busted.outputHandlers.${BUSTED_OUTPUT_TYPE}
```

Comment (justinmk) [inline cmake/RunTests.cmake:78]: i see `-ll` was used in the old code but should it be `-l` ?

Comment (bfredl) [reply cmake/RunTests.cmake:78]: should be `-ll` as is. we want full control of initialization in tests.

File: test/client/session.lua [hunk 2]
```diff
@@ -0,0 +1,194 @@
+local safe_pcall
```
Comment (justinmk) [inline test/client/session.lua:1]: is `test/client/session.lua` actually only for tests, or (eventually) would we move it to `runtime/...` ?

Comment (bfredl) [reply test/client/session.lua:1]: eventually. regardless, `coxpcall` is very small and pure lua, so we could vendor it in `runtime/` when this happens.

File: test/functional/helpers.lua [hunk 3]
```diff
@@ -1,14 +1,12 @@
...
 local lfs = require('lfs')
-local mpack = require('mpack')
 local global_helpers = require('test.helpers')
 
--- nvim client: Found in .deps/usr/share/lua/<version>/nvim/ if "bundled".
-local Session = require('nvim.session')
-local TcpStream = require('nvim.tcp_stream')
-local SocketStream = require('nvim.socket_stream')
-local ChildProcessStream = require('nvim.child_process_stream')
+-- nvim client: Found in runtime/vim/client :)
```
Comment (zeertzjq) [inline test/functional/helpers.lua:5]:
```suggestion
-- nvim client: Found in test/client :)
```

File: test/client/uv_stream.lua [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+  end
+  self._closed = true
+  self:read_stop()
+  self._child_stdin:close()
+  self._child_stdout:close()
+  if type(signal) == 'string' then
+    self._proc:kill('sig'..signal)
+  end
+  uv.run('nowait')
+end
```
Comment (zeertzjq) [inline test/client/uv_stream.lua:158]: Does this no longer wait for the process to exit? Then some warnings about session taking too long to exit in the test suite may no longer make sense.

Comment (zeertzjq) [reply test/client/uv_stream.lua:158]: It's possible to use `expect_exit(command, 'qall!')` to make sure the process exits, but that may be a bit slower as it makes an RPC request.

Comment (bfredl) [reply test/client/uv_stream.lua:158]: we could add a wait for exit the proper way (run loop until exit callback, with a timeout)

**Pull request merged**